File: javascript/9 - grafos.js
class Dicionario {
  constructor(){
    this.items = {}
  }

  set(key, value){
    this.items[key] = value
  }

  delete(key) {
    if(this.has(key)) {
      delete this.items[key]
      return true
    }
    
    return false
  }

  has(key) {
    return this.items.hasOwnProperty(key)
  }

  get(key) {
    return this.has(key) ? this.items[key] : null 
  }

  clear(){
    this.items = {}
  }

  size(){
    return Object.keys(items).length
  }

  keys() {
    return Object.keys(items)
  }

  value(){
    return Objexct.values(items)
  }

  getItems(){
    return this.items
  }
}

class Fila {
  constructor() {
    this.fila = [];
  }

  adicionar(elemento) {
    this.fila.push(elemento);
  }

  remover() {
    return this.fila.shift();
  }

  vazia() {
    return this.fila.length === 0;
  }

  proximo() {
    return this.fila[0]
  }

  limpar() {
    this.fila = []
  }

  imprimir()  {
      console.log(this.fila)
  }
}

class Grafos {
  constructor(){
    this.vertices = []
    this.adjList = new Dicionario()
  }

  addVertex(vertice){
    this.vertices.push(vertice)
    this.adjList.set(vertice, [])
  }

  addEdge(vertice, edge){
    this.adjList.get(vertice).push(edge)
    this.adjList.get(edge).push(vertice)
  }

  toString(){
    let str = ''
    for(let i = 0; i < this.vertices.length; i++){
      str += this.vertices[i] + '->'
      var neigbhores = this.adjList.get(this.vertices[i])
      for(let j = 0; j < neigbhores.length; j++){
        str += neigbhores[j] + ' '
      }
      str += '\n'
    }
    return str
  }

  initializeColor(){
    const color = []
    for(let i = 0; i < this.vertices.length; i++){
      color[vertices[i]] = 'white'
    }
    
    return color
  }

  buscaEmProfundidadeDFS(vertice, callback) {
    const color = this.initializeColor()
    for(var i = 0; i < this.vertices.length; i++){
      if(color[vertices[i]] === 'white') {
        this.dfsVisit(vertices[i], color, callback)
      }
    }
  }

  dfsVisit(vertice, color, callback){
    color[vertice] = 'gray'
    callback(vertice)
    var neigbhores = this.adjList.get(vertice)
    for(let i = 0; i < neigbhores.length; i++){
      var w = neigbhores[i]
      if(color[w] === 'white'){
        this.dfsVisit(w, color, callback)
      }
    }
    color[vertice] = 'black'
  }

  buscaEmLarguraBFS(vertice, callback){
    const color = this.initializeColor()
    const queue = new Fila()
    queue.adicionar(vertice)

    while(!queue.vazia()){
      var u = queue.remover()
      neigbhores = this.adjList.get(u)
      color[u] = 'gray'
      for(let i = 0; i < neigbhores.length; i++){
        var w = neigbhores[i]
        if(color[w] === 'white'){
          color[w] = 'gray'
          queue.adicionar(w)
        }
      }
      color[u] = 'black'
      callback(u)
    }
  }
}


function visitedVertice(v){
  console.log(v)
}

File: ruby/9 - grafos.rb
class Dicionario
  attr_reader :items

  def initialize
    @items = {}
  end

  def set(key, value)
    @items[key] = value
  end

  def delete(key)
    @items.delete(key)
  end

  def has(key)
    @items.has_key?(key)
  end

  def clear
    @items.clear
  end

  def size
    @items.size
  end

  def keys
    @items.keys
  end

  def values
    @items.values
  end
end

class Fila
  attr_reader :fila

  def initialize()
    @fila = []
  end

  def adicionar(value)
    fila.push(value)
  end

  def remover
    fila.shift
  end

  def proximo
    fila.first
  end

  def vazia
    fila.empty?
  end

  def limpar
    file.clear
  end

  def tamanho
    fila.size
  end

  def imprimir
    puts "Fila: #{fila}"
  end
end

class Grafos
  def initialize
    @vertices = []
    @arestas = Dicionario.new  
  end

  def adicionar_vertice(vertice)
    @vertices.push(vertice)
    @arestas.set(vertice, [])
  end

  def adicionar_aresta(vertice, edge)
    @arestas.get(vertice).push(edge)
    @arestas.get(edge).push(vertice)
  end

  def to_s
    str = ""
    for(i =0; i < @vertices.size; i += 1)
      str += @vertices[i].to_s + ": ->"
      visinhos = @arestas.get(@vertices[i])
      for(j=0; j < visinhos.size; i += 1)
        str += visinhos[j].to_s + " "
      end 
      str += "\n"
    end
    str
  end

  def inicializar_cor
    cor = []
    for(i = 0; i < @vertices.size; i += 1)
      cor[@vertices[i]] = "white"]
    end
  
    cor
  end

  def busca_em_profundidadeDFS(block)
    colors = inicializar_cor
    @vertices.each_with_index do |vertice, index|
      if(colors[vertice[index]] == "white")
        dfs_visit(vertice, colors, block)
      end
    end
  end

  def bfs_busca_em_largura(vertice, block)
    colors = initializar_cor
    fila = Fila.new
    fila.adicionar(vertice)

    while(!fila.vazia)
      u = queue.remover
      neighors = @arestas.get(u)
      color[u] = "gray"
      neighors.each_with_index do |value, index|
        if(colors[value] == "white")
          colors[value] = "gray"
          fila.adicionar(value)
        end 
      end
      color[u] = "black"
      block(u)
    end
  end

  private

  def dfs_visit(vertice, colors, block)
    colors[vertice] = "gray"
    block.call(vertice)
    visinhos = @arestas.get(vertice)
    visinhos.each_with_index do |visinho, index|
      w = neighors[index]
      if colors[w] == "white"
        color[w] = "gray"
        dfs_visit(w, colors, block)
      end
    end
  
    colors[vertice] = "black"
  end
end

def visitedVertices(vertice)
  print vertice
end